crates/c-api/include/wasi.h

```diff
@@ -0,0 +1,70 @@
+// WASI C API
+
+#ifndef WASI_H
+#define WASI_H
+
+#include "wasm.h"
+
+#ifndef WASI_API_EXTERN
+#ifdef _WIN32
+#define WASI_API_EXTERN __declspec(dllimport)
+#else
+#define WASI_API_EXTERN
+#endif
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define own
+
+#define WASI_DECLARE_OWN(name) \
+  typedef struct wasi_##name##_t wasi_##name##_t; \
+  WASI_API_EXTERN void wasi_##name##_delete(own wasi_##name##_t*);
+
+// WASI config
+
+WASI_DECLARE_OWN(config)
+
+WASI_API_EXTERN own wasi_config_t* wasi_config_new();
+
+WASI_API_EXTERN void wasi_config_set_argv(wasi_config_t* config, int argc, const char* argv[]);
+WASI_API_EXTERN void wasi_config_inherit_argv(wasi_config_t* config);
+
+WASI_API_EXTERN void wasi_config_set_env(wasi_config_t* config, int envc, const char* names[], const char* values[]);
+WASI_API_EXTERN void wasi_config_inherit_env(wasi_config_t* config);
+
+WASI_API_EXTERN bool wasi_config_set_stdin(wasi_config_t* config, const char* path);
+WASI_API_EXTERN void wasi_config_inherit_stdin(wasi_config_t* config);
+
+WASI_API_EXTERN bool wasi_config_set_stdout(wasi_config_t* config, const char* path);
```

Comment:
It would be useful to have support for input and output streams to hook stdin/stdout/stderr without having to replace the file handles which gets tricky when these are used in the same process. Having the stream support would make it simple to support multiple WASI instances in the same process.

WASI_API_EXTERN bool wasi_config_set_stdin(wasi_config_t* config, wasi_input_stream_t* stream, void* env);
WASI_API_EXTERN bool wasi_config_set_stdout(wasi_config_t* config, wasi_output_stream_t* stream, void* env);
WASI_API_EXTERN bool wasi_config_set_stderr(wasi_config_t* config, wasi_output_stream_t* stream, void* env);

env is a pointer forwarded to the input/output stream.

The caller would then just need to implement the 2 binders to hook the stdio.

Reply:
Thanks for the feedback!  Indeed that would be desirable.  This is only a first stab at the WASI API and it was just modeled after what `WasiCtxBuilder` supports.  I think a proper stream abstraction would make a lot of sense here and in the Rust API.  I think that can be added after this lands, though.

Reply:
The file path in stdio set methods looks like a temporary solution (in favor of pipe or stream), can we rename these so "good" names will be available for the stream/pipes?

Reply:
Can do.  Maybe `wasi_config_set_std[in|out|err]_file`?

Reply:
I've renamed to `wasi_config_set_std[in|out|err]_file` so we can keep the old name for when we properly support streams.